driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java

```diff
@@ -710,6 +759,7 @@ public void onSet(
             retryCount,
             queryState,
             queryStateRef.get());
+        currentTracingInfo.tracingFinished();
```

Comment:
Regarding `currentTracingInfo.tracingFinished()` in `onSet()`: I believe that all those invocations could be moved to a single `tracingFinished()` after `currentTracingInfo.setShardID(connection.shardId())`. The downside is that technically it would be done a bit too early (not waiting for `connection.release()`), but it would remove the possibility that someone forgets to put `tracingFinished()` in some branch.

Reply:
Yeah, I did it this way because I was wondering what to do in case of getting response `UNPREPARED`. Now in this case the tracing isn't finished in `onSet()`, but later in `prepareAndRetry()` callback, cause `write(connection, prepareAndRetry(toPrepare.getQueryString()))` uses the same connection. But creating a new span for this write and ending it in `prepareAndRetry()` may be a better option.